stablehlo/dialect/TypeInference.h

```diff
@@ -78,6 +78,10 @@ LogicalResult verifyReducerShape(Optional<Location> loc, Block& block,
                                  ArrayRef<int64_t> allowedDimensions,
                                  bool allInputsUnranked);
 
+LogicalResult verifyReduceScatter(Operation* op, TypeRange operandTypes,
+                                  TypeRange resultTypes,
+                                  uint64_t scatterDimension);
```

Comment:
This function is called exactly once in `verifyReduceScatterOp`. Let's inline it into the callsite?

Reply:
I kept it because it also declared (probably reused) in https://github.com/tensorflow/mlir-hlo/blob/master/mhlo/IR/hlo_ops_common.h, and then used by `https://github.com/tensorflow/mlir-hlo/blob/greencommit/2022-12-19-7ccbb4df/lhlo/IR/lhlo_ops.cc#L149`.  So I assume it's better to keep it as shared method?

Reply:
Sounds like the linked code in LMHLO should be using `verifyReduceScatterOp` directly instead of `verifyReduceScatter`. Let's kill `verifyReduceScatter` in this PR and then propagate that to MLIR-HLO in a StableHLO to MHLO ticket.

Reply:
I'm happy to change but there're some detail misaligned:
1. `ReduceScatterOp` in LMHLO has different  arguments like `inputs/outputs`. Now I'm wrapping everything from  `ReduceScatterOp::verify()` in StableHLO to the new `verifyReduceScatterOp(..., Value operand, Value result)`, which does not fit for LMHLO
2. Except for the common calls to `verifyReduceScatter()`, other parts are different: MHLO have more verifications and I prefer not to include complicated change in this large PR.

StableHLO verifier:
https://github.com/openxla/stablehlo/blob/main/stablehlo/dialect/StablehloOps.cpp#L280
```
LogicalResult ReduceScatterOp::verify() {
  if (failed(hlo::verifyReplicaGroups(getLoc(), getReplicaGroups(),
                                      /*allGroupsMustHaveSameSize=*/true,
                                      getUseGlobalDeviceIds(),
                                      /*expectedGroupSize=*/std::nullopt)))
    return failure();
  auto operandType = getOperand().getType().cast<TensorType>();
  bool operandTypeRanked = operandType.isa<RankedTensorType>();
  Block& block = getComputation().front();
  if (failed(hlo::verifyReducerShape(
          this->getLoc(), block, {operandType},
          {RankedTensorType::get({}, operandType.getElementType())},
          /*numInputs=*/1, /*allowedDimensions=*/{},
          /*allInputsUnranked=*/!operandTypeRanked)))
    return failure();

  return verifyReduceScatter(*this,
                             /*operandTypes=*/{getOperand().getType()},
                             /*resultTypes=*/{getType()},
                             /*scatterDimension=*/getScatterDimension());
}
```

LMHLO verifier:
https://github.com/tensorflow/mlir-hlo/blob/greencommit/2022-12-19-7ccbb4df/lhlo/IR/lhlo_ops.cc#L149
```
LogicalResult ReduceScatterOp::verify() {
  ReduceScatterOp op = *this;
  if (failed(mlir::hlo::verifyReplicaGroups(op, /*isUniformSized=*/true)))
    return failure();
  if (failed(mlir::hlo::verifyReduceScatter(
          op, /*operandTypes=*/op.getInputs().getTypes(),
          /*resultTypes=*/op.getOutputs().getTypes(),
          /*scatterDimension=*/op.getScatterDimension())))
    return failure();
  return success();
}
```

What do you think of it?

Reply:
Let's inline the function in this PR and have a separate ticket for code sharing with LMHLO.

Reply:
Sure, with a few changes:
1. The big for loop is not necessary any more: it only loop once with operand/result instead of operands/results
2. Some "continue" are replaced with "return success()"
3. Some verifications moved together.